File: dotnet/src/Grafana.Agento11y/SdkVersion.cs
using System.Reflection;

namespace Grafana.Agento11y;

/// <summary>
/// SDK version and User-Agent product token.
/// </summary>
/// <remarks>
/// <see cref="Version"/> is stamped into the default generation-export User-Agent
/// (see <see cref="UserAgent"/>). Resolved from the assembly's informational version;
/// <c>0.0.0+unknown</c> when that metadata is unavailable.
/// </remarks>
public static class SdkVersion
{
    /// <summary>Version of the Agento11y .NET SDK.</summary>
    public static readonly string Version = ResolveVersion();

    private const string UserAgentProduct = "agento11y-sdk-dotnet";

    private static string ResolveVersion()
    {
        var informational = typeof(SdkVersion).Assembly
            .GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
        if (informational == null || string.IsNullOrWhiteSpace(informational))
        {
            return "0.0.0+unknown";
        }

        // SourceLink/ContinuousIntegrationBuild appends "+<sha>"; strip it.
        var plus = informational.IndexOf('+');
        return plus > 0 ? informational.Substring(0, plus) : informational;
    }

    /// <summary>
    /// Returns the SDK's default generation-export User-Agent product token,
    /// <c>agento11y-sdk-dotnet/&lt;Version&gt;</c>.
    /// </summary>
    public static string UserAgent() => UserAgentProduct + "/" + Version;
}

File: java/core/src/main/java/com/grafana/agento11y/sdk/SdkVersion.java
package com.grafana.agento11y.sdk;

/**
 * SDK version and User-Agent product token.
 *
 * <p>{@link #VERSION} is stamped into the default generation-export User-Agent (see {@link
 * #userAgent()}). Resolved from the jar manifest's {@code Implementation-Version};
 * {@code 0.0.0+unknown} when manifest metadata is unavailable (e.g. classes-directory runs).
 */
public final class SdkVersion {
    /** Version of the Agento11y Java SDK. */
    public static final String VERSION = resolveVersion();

    private static final String USER_AGENT_PRODUCT = "agento11y-sdk-java";

    private SdkVersion() {}

    private static String resolveVersion() {
        String version = SdkVersion.class.getPackage().getImplementationVersion();
        return (version == null || version.isBlank()) ? "0.0.0+unknown" : version;
    }

    /**
     * Returns the SDK's default generation-export User-Agent product token, {@code
     * agento11y-sdk-java/<VERSION>}.
     */
    public static String userAgent() {
        return USER_AGENT_PRODUCT + "/" + VERSION;
    }
}
